### PR TITLE
Anti-stutter work

### PR DESCRIPTION
**Pull request by jherico**
This is a first pass at reducing the amount of stutter caused when entering a large scene.  

The stutter appears to be caused by a number of factors.  

- overloading the GPU with an excessive number of simultaneous transfers of large amount of texture / vertex data
- overloading the CPU with an excessive number of parsing threads doing file IO

This PR attempts to address the latter component first, simply because it's easier.  It reduces the size of the global thread pool from the default of `QThread::idealThreadCount()` to `2`.  This means that if you have 16 textures/models to load to render a scene, only 2 of them will be concurrently being loaded from disk to CPU memory at a time.  

Previously it would allow all CPU cores to be consumed performing this task, which could starve the main thread and the present thread, preventing keyboard or mouse input from being handled, and causing the screen to simply freeze (regardless of threaded present).  

Because the number of concurrent file transfers is being limited, the GPU transfers are also partially serialized, reducing their impact as well.

As a result, when teleporting from a nearly empty domain to `hifi://demo/547.046,495.032,509.714/0,-0.29785,0,-0.954613`, the number of missed frames (caused by the main / rendering thread being blocked/starved) is reduced and the amount of complete freezes (caused by the present thread being starved) is almost completely eliminated.

### CODE REVIEW

**Comment by jherico** [comment]
This is a first pass for work on https://app.asana.com/0/27650181942747/86757725244686

**Comment by hifi-gustavo** [comment]
Build console output for commit: 59838587cf62a400053f5d726da13ca60d5acfc4 available at: https://gist.github.com/6cb4840bbfe695e3b468<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7049.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7049.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7158a2978edad4f84c3e0dc5f072d988105bb40d available at: https://gist.github.com/6cb4840bbfe695e3b468<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7049.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7049.dmg

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -4762,8 +4774,11 @@ void Application::updateDisplayMode() {
         foreach(auto displayPlugin, standard) {
             addDisplayPluginToMenu(displayPlugin, first);
             // This must be a queued connection to avoid a deadlock
-            QObject::connect(displayPlugin.get(), &DisplayPlugin::requestRender,
-                this, &Application::paintGL, Qt::QueuedConnection);
+            QObject::connect(displayPlugin.get(), &DisplayPlugin::requestRender, [=] {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: Why change this from a Qt::QueuedConnection to a direct connection with a postEvent()? What is the effect of that change? Is it intended to address stuttering? How? Does it have no other side effects? How did you verify that?

File: interface/src/ui/ApplicationCompositor.cpp [hunk 2]
```diff
@@ -286,6 +287,24 @@ void ApplicationCompositor::displayOverlayTextureHmd(RenderArgs* renderArgs, int
...
+        // with the window.  This means if it's being driven by the mouse and...
+        // * The mouse is not over the Interface window
+        // * The interface window is not active
+        // However, if we're using something else to drive the pointer (hand controllers)
+        // then we don't care where the mouse cursor is.  However, interacting with the 
+        // QML UI probably still won't work if the interface window is not active/focused
+        // because the QML offscreen rendering uses the actual window as a proxy for detecting
+        // focus.
+        
+        // Don't render the mouse cursor if we aren't the active window
```
Comment (ZappoMan) [inline interface/src/ui/ApplicationCompositor.cpp]: I'm not convinced this is the right thing to do. And we certainly shouldn't be doing it in this PR... this change obviously has nothing to do with stuttering, and violates the "fix reported bugs" rule. Please remove this from this PR. We can file a different bug about mouse cursor/focus and active window, and we can discuss the correct fix in that context. For example, I am convinced we probably need to change our application to FORCE always on top when in HMD mode (or something similar)... but we don't want to do that when we're in normal desktop mode.

The point is... no... this "fix" is not correct. This is a merge blocker.

File: libraries/animation/src/AnimationCache.cpp [hunk 3]
```diff
@@ -47,6 +47,10 @@ AnimationReader::AnimationReader(const QUrl& url, const QByteArray& data) :
 }
 
 void AnimationReader::run() {
+    auto originalPriority = QThread::currentThread()->priority();
+    if (originalPriority == QThread::InheritPriority) {
+        originalPriority = QThread::NormalPriority;
+    }
```
Comment (ZappoMan) [inline libraries/animation/src/AnimationCache.cpp]: Did you miss a call to `QThread::currentThread()->setPriority(QThread::LowPriority);`? If you didn't intend to change the priority here, why are you storing the original priority?

File: libraries/model-networking/src/model-networking/ModelCache.cpp [hunk 4]
```diff
@@ -50,6 +50,11 @@ GeometryReader::GeometryReader(const QUrl& url, const QByteArray& data, const QV
 }
 
 void GeometryReader::run() {
+    auto originalPriority = QThread::currentThread()->priority();
+    if (originalPriority == QThread::InheritPriority) {
```
Comment (ZappoMan) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: it seems as though this code can't/shouldn't ever actually execute. AFAICT... `priority()` will only return `QThread::InheritPriority` if the thread in question is not running... but if that was the case, then `setPriority(x);` also "does nothing". Also, since you're in the `run()` method we pretty much can assume your thread has started. So... I don't think we need to check here, and if anything we probably want to simply `assert(originalPriority !== QThread::InheritPriority)`

File: libraries/model/src/model/TextureMap.cpp [hunk 5]
```diff
@@ -50,88 +50,55 @@ void TextureMap::setLightmapOffsetScale(float offset, float scale) {
 }
 
 
-
-
+// FIXME why is this in the model library?  Move to GPU or GPU_GL
 gpu::Texture* TextureUsage::create2DTextureFromImage(const QImage& srcImage, const std::string& srcImageName) {
     QImage image = srcImage;
- 
-    int imageArea = image.width() * image.height();
```
Comment (ZappoMan) [inline libraries/model/src/model/TextureMap.cpp]: why are you removing all this code? This doesn't seem to belong in this PR... if there's some other bug you're fixing the make this a different PR, with a clear asana bug report about what you're fixing.

Please remove this change.

Comment (ZappoMan) [inline libraries/model/src/model/TextureMap.cpp]: I now see you named the commit as "don't burn CPU cycles calculating average color we don't use"... it would make my review easier if you had included that comment in the PR so I knew to look for the change... and inline comment here would be useful.

That being said, it looks like you also changed the logic a bit. And/or there's multiple calls to `image = image.convertToFormat(QImage::Format_RGB888);` that likely can be consolidated.

File: interface/src/Application.cpp (hunk 1 above)
Comment (jherico) [inline interface/src/Application.cpp]: Both versions of the function do the same thing, asynchronously execute `paintGL`.  However, the old version just puts the triggering signal onto a stack of signals to be processed.  The new version ensures that the asynchronous posted message will have higher priority that items already in the queue with normal priority.  

This should result in less chance of a frame being missed if the event queue is being saturated or the main thread is being starved of CPU usage.

File: interface/src/ui/ApplicationCompositor.cpp (hunk 2 above)
Comment (jherico) [inline interface/src/ui/ApplicationCompositor.cpp]: well, OK, I'm just really tired of trying to click things in interface and they don't work, and I finally take off my headset and discover I've ordered 2 gross of imitation dutch wooden shoes from china.  I will rebase the commit out of the PR.

**Comment by OmegaHeron** [comment]
Tested on Windows 7.  Compared to current production build this is an improvment - I used hifi://sandbox as my comparison point, teleporting from a low impact scene to it.  With this PR AV is (somewhat) responsive during loading, overall scene rez seems smoother.  With production code AV is totally unresponsive during load and scene builds in a chaotic/jarring fashion.  This seems like a decent start to better things.

File: libraries/model-networking/src/model-networking/ModelCache.cpp (hunk 4 above)
Comment (jherico) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: Before I put in this clause I kept getting log messages that said `You cannot set thread priority to InheritPriority`

File: libraries/model/src/model/TextureMap.cpp (hunk 5 above)
Comment (jherico) [inline libraries/model/src/model/TextureMap.cpp]: There were always multiple calls to `convertToFormat(QImage::Format_RGB888);`, however I think I can consolidate.

File: libraries/animation/src/AnimationCache.cpp (hunk 3 above)
Comment (jherico) [inline libraries/animation/src/AnimationCache.cpp]: fixed

File: libraries/model/src/model/TextureMap.cpp [hunk 6]
```diff
@@ -50,88 +50,55 @@ void TextureMap::setLightmapOffsetScale(float offset, float scale) {
 }
 
 
-
-
+// FIXME why is this in the model library?  Move to GPU or GPU_GL
```
Comment (samcake) [inline libraries/model/src/model/TextureMap.cpp]: THis actually belongs here not in GPU
THe goal is to take care of the different variations on how we expect the textures representing the material properties to be created and recombined
This is on the asset side hence here.

File: libraries/model/src/model/TextureMap.cpp [hunk 7]
```diff
@@ -50,88 +50,55 @@ void TextureMap::setLightmapOffsetScale(float offset, float scale) {
...
 gpu::Texture* TextureUsage::create2DTextureFromImage(const QImage& srcImage, const std::string& srcImageName) {
     QImage image = srcImage;
- 
-    int imageArea = image.width() * image.height();
-
```
Comment (samcake) [inline libraries/model/src/model/TextureMap.cpp]: WE don't use it yet but the goal of that code is to differentiate between the mask alpha vs true transparency channel and classify the material accordingly (opaque/texel opaque /transparent)
We can probably consolidate the the 2 branches of the the if, but we need do need this to happen there

File: libraries/animation/src/AnimationCache.cpp (hunk 3 above)
Comment (ZappoMan) [inline libraries/animation/src/AnimationCache.cpp]: you appear to not have yet pushed your fix... at least github doesn't have the change.

File: libraries/model/src/model/TextureMap.cpp (hunk 6 above)
Comment (jherico) [inline libraries/model/src/model/TextureMap.cpp]: Maybe the format conversion code belongs here, but it seems like the the actual grunt work of taking a QImage and turning it into a gpu::Texture should be in GPU.

File: libraries/model/src/model/TextureMap.cpp (hunk 7 above)
Comment (jherico) [inline libraries/model/src/model/TextureMap.cpp]: We can restore the code once we're actually using it, presumably with a better implementation than calling `pixel()` for every actual pixel in the image, which is the absolute worst performing way of doing this.  

Right now this code both poorly optimized *and* produces no output.  I'm not going to optimize it if it's not used, so... poof.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7d99f9e72f574aee8bc88f86bd98f7dbafec3fbb available at: https://gist.github.com/6cb4840bbfe695e3b468<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7049.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7049.dmg

**Comment by OmegaHeron** [comment]
Retested on Windows 7 with latest build - same eval as before, an improvement over current production code for loading a complex scene.

File: libraries/model-networking/src/model-networking/ModelCache.cpp (hunk 4 above)
Comment (ZappoMan) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: https://www.youtube.com/watch?v=0tJGk4ofc18
